sphinx/domains/c.py

```diff
@@ -271,65 +271,72 @@ class CDomain(Domain):
     }  # type: Dict[str, Dict[str, Tuple[str, Any]]]
 
     @property
-    def objects(self) -> Dict[str, Tuple[str, str]]:
-        return self.data.setdefault('objects', {})  # fullname -> docname, objtype
+    def objects(self) -> Dict[Tuple[str, str], str]:
+        return self.data.setdefault('objects', {})  # objtype, fullname -> docname
```

Comment:
I'll change the structure of `self.objects` later to add node_id into the value in next PR. It is also needed to avoid conflicts with other domains (refs: #6903).